### ISSUE TITLE
[Feature] LSP DebugInfo request

### ISSUE THREAD

**Issue by kbobyrev** [Microsoft/language-server-protocol#558]
For debugging and troubleshooting purposes, it is extremely useful to have debug information printer in the editor. Most Vim clients either don't provide any way of printing out LSP server address, log file, etc or use their own format (e.g. I have to [ask](https://github.com/autozimu/LanguageClient-neovim/issues/558) developers to add this feature). It might be useful for both users and LSP developers to provide a unified request (`workspace/debugInfo`?) and debug information format. As an LSP server developer, I'm mainly interested in

* Server PID (so that I could attach my debugger to the process)
* Log files: both `stdout` and `sterr` files
* URL
* (maybe) root of the project

Having this information would provide better user experience for bug submissions and better developer experience.

This is an example of debug information printer in Vim's [YouCompleteMe](https://github.com/Valloric/YouCompleteMe):

![43989813-6e98d63a-9d51-11e8-93a1-62125c874c88](https://user-images.githubusercontent.com/3352968/44705455-f6fb0200-aa9e-11e8-8628-ca5f5046792d.png)

**Comment by kbobyrev** [comment]
Thanks for the feedback!

> Moreover, debugging can happen even before LS manages to start the connection (cannot connect to client or server, erroneous startup...), so in those cases (which are IMO the most frequent reasons to debug a LS), the operation wouldn't even be accessible.

Sorry, the wording might not have been great: we probably mean different things by "debugging" then. My use-case actually splits into monitoring the LS behavior (e.g. how much memory it consumes) and reproducing the problems (e.g. attaching to the LS process, inspecting the internal state, etc). I think users would benefit from both and I don't think that dumping debug output to the log file is sufficient for that purpose.

I'll rephrase in the original request.

**Comment by kbobyrev** [comment]
@LaurentTreguier yes, I agree. Also, for some languages (e.g. C++) it's really hard to guess where the root is (and what it actually is). I'll update the initial comment.

Yes, server version could go into the additional field which can specific for each server.

**Comment by dbaeumer** [comment]
Just to be sure I interpret this correctly. This would be a notification send from the client to the server to turn on some tracing / logging. The server would then use the notification available to write to the console of the client.

**Comment by kbobyrev** [comment]
My initial proposal was the server info pretty-printing request: client sends request to show information about the server which is currently used to see where it stores logs, traces which version it has and other additional server information. Turning on logging/tracing can be done via passing flags to the server, that's probably something different.

**Comment by robertoaloi** [comment]
Related to this: version `3.15` of the protocol introduces `ServerInfo` as part of the `InitializationResult`. Maybe we could clarify how clients should use that information. It could also make sense to introduce a `request` from the client to the server to request "serverinfo". In that case the format should be relaxed a bit (since right now only contains `name` and `version`). Maybe also return a property list?
